### DISCUSSION TITLE
Rate storage in DB as rounded values?

### DISCUSSION THREAD

**Discussion by Cayburn** [kimai/kimai#3539] [Questions & Answers]
It seems that the rate is stored in the DB as rounded values. When we have an hourly rate of 99,75 and have a duration of 30 minutes, the rate is stored as 49,88 and not 49,875. This causes issues on the invoices when the sums are created. Is there an option to store the rates with full precision? I updated rate in the DB directly on my test system and it seems to work perfectly fine.

**Comment by kevinpapst** [comment]
How many decimals are full precision for you?

**Comment by Cayburn** [comment]
I would select 6 digits to be on the save side for middle-huge sized companies.

That depends on the number of expected time entries. Assuming 2 decimals, each time entry might differ by 0.5 cent Having already 3 entries might have rounding issues. Using 3 decimals, each time entry might differ by 0.05 cent - 20+1 entries might cause issues. These are theoretical worst case issues 0.49999 is tough to round to.

- 2 decimals -> 2 entries with might cause issues
- 3 decimals -> 21 entries
- 4 decimals -> 201 entries
- 5 decimals -> 2001 entries
- 6 decimals -> 20001 entries

**Reply by kevinpapst** [reply]
Again: interesting! Thanks for sharing.

I have discussed that often and I have all kind of customers from small to large creating invoices per month with Kimai. 
Can you please explain in detail what is your issue.

**Reply by Cayburn** [reply]
The problem is that we are using cent amounts for hourly rates and have quarterly ranged time entries. On the invoice the "rate" is summed up, so when one entry is having 1 cent less, the the total is not what is expected. In our prod system we had a difference of 36 Cents. On the test system I used 12 time entries (15', 30', 45', 1:15') for an hourly rate or 114,75 and that summed up for 10 hours not as expected 1.147,49 € (on expected 1.147,50 €)

![image](https://user-images.githubusercontent.com/13998916/190391391-390f1d3b-7c96-4ae4-ba04-2281bb0a829b.png)

**Reply by kevinpapst** [reply]
Are you up for a PR including tests?

**Reply by Cayburn** [reply]
Hi Kevin, unluckily I am not familar enough with Synfony and have no experiences at all with the testing framework.

I would implement a new setting "rate rounding factor" with a default of 2 and possible values for 2-10 and would then apply this at the place where it's rounded. Currently I am using a direct SQL in the database which is not really elegant.

**Reply by kevinpapst** [reply]
I created an issue to track that topic: https://github.com/kevinpapst/kimai2/issues/3592
